sql/src/main/java/io/crate/action/sql/SQLOperations.java

```diff
@@ -265,28 +288,28 @@ public void bind(String portalName, String statementName, List<Object> params, @
 
             try {
                 Portal portal = getSafePortal(portalOrStatement);
-                Analysis analysis = portal.analysis;
+                Analysis analysis = portal.getLastAnalysis();
                 if (analysis.rootRelation() == null) {
                     return null;
                 }
                 List<Field> fields = analysis.rootRelation().fields();
-                portal.outputTypes = Symbols.extractTypes(fields);
+                portal.addOutputTypes(Symbols.extractTypes(fields));
                 return fields;
             } catch (Throwable t) {
                 throwable = t;
                 throw t;
             }
         }
 
-        public void execute(String portalName, int maxRows, ResultReceiver rowReceiver) {
+        public void execute(String portalName, int maxRows, ResultReceiver resultReceiver) {
             LOGGER.debug("method=execute portalName={} maxRows={}", portalName, maxRows);
 
             checkError();
             try {
                 Portal portal = getSafePortal(portalName);
-                validateReadOnly(portal.analysis);
-                portal.addRowReceiver(rowReceiver);
-                portal.maxRows = maxRows;
+                validateReadOnly(portal.getLastAnalysis());
+                portal.addResultReceiver(resultReceiver);
```

Comment:
same here, maybe introduce some kind of `prepareExecute` on the portal as you already mentioned. Then you can require all the params in one method